### PR TITLE
Avoid GDScript bookkeeping from referencing objects longer than necessary

### LINKED ISSUE #66818: `FileAccess` produces `.tmp` file when used inside `EditorPlugin._build` callback.
### Godot version

v4.0.beta2.official [f8745f2f7]

### System information

Windows 10

### Issue description

When writing to a file in the `_build` callback of an `EditorPlugin` and afterwards printing the list of files in this directory, the file is not in the list. Instead a file with `.tmp` added is contained. The following code was used:
```swift
@tool
extends EditorPlugin


func _build() -> bool:
	var file = FileAccess.open("res://test.txt", FileAccess.WRITE)
	file.store_string("test")
	file.flush()
	print(DirAccess.get_files_at("res://"))
	return false
```
Output: `[".gitattributes", ".gitignore", "icon.svg", "icon.svg.import", "project.godot", "test.txt.tmp"]`

Using the old API's in beta1 this did not occure.
Code in beta1:
```swift
@tool
extends EditorPlugin


func _build() -> bool:
	var file = File.new()
	file.open("res://test.txt", File.WRITE)
	file.store_string("test")
	file.close()

	var dir = Directory.new()
	dir.open("res://")
	print(dir.get_files())
	return false

```
Output (beta1): `[".gitattributes", ".gitignore", "icon.svg", "icon.svg.import", "project.godot", "test.txt"]`

At first I thought it could have something to do with the removed `close` method. But the beta2 code works as expected, when used in a normal script. So this must be some edge case.

### Steps to reproduce

- Create an editor plugin with the code from above.
- Try to run in beta1 and beta2
- Observe the output

### Minimal reproduction project

[file_access_bug.zip](https://github.com/godotengine/godot/files/9697314/file_access_bug.zip)

### LINKED ISSUE #68614: ZIPReader with HTTPRequest doesn't work consistently in seemingly equivalent code
### Godot version

4.0 beta 4

### System information

Windows 10

### Issue description

I am working on some code to download a zip file using HTTPRequest then reading it with the new ZIPReader. In the example issue attached, I have 3 scripts attached to 3 buttons that, as far as I understand should work in an equivalent fashion. However only one of them appears to successfully read the zip file.

Button 1, the one that works just tucks everything into the function connected to request_completed signal, here we succesfully reach the "I'm good" message

```gdscript
extends Button

var download_path = 'down/zip.zip'
var download_link = 'https://github.com/oganm/ZIPReader_issue/archive/refs/heads/main.zip'

func _ready():
    pressed.connect(self._on_pressed)

func _on_pressed():
    DirAccess.remove_absolute(download_path)
    var http_request = HTTPRequest.new()
    add_child(http_request)
    http_request.request_completed.connect(self._download_completed)
    http_request.request(download_link)

func _download_completed(result, response_code, headers, body):
    var install_directory = download_path.get_base_dir()
    DirAccess.make_dir_recursive_absolute(install_directory)

    var f:FileAccess = FileAccess.open(download_path,FileAccess.WRITE)
    f.store_buffer(body)
    f = null

    var reader = ZIPReader.new()
    var err = reader.open(download_path)
    print(err)
    if err != OK:
        push_error('An error occured reading the zip file')
        return err
    else:
        print("I'm good")
```
The only difference of Button 2 is that the code for reading the zip is moved to a different function but this fails.
```gdscript
extends Button

var download_path = 'down/zip.zip'
var download_link = 'https://github.com/oganm/ZIPReader_issue/archive/refs/heads/main.zip'

func _ready():
    pressed.connect(self._on_pressed)

func _on_pressed():
    DirAccess.remove_absolute(download_path)

    var http_request = HTTPRequest.new()
    add_child(http_request)
    http_request.request_completed.connect(self._download_completed)
    http_request.request(download_link)

func _download_completed(result, response_code, headers, body):
    var install_directory = download_path.get_base_dir()
    DirAccess.make_dir_recursive_absolute(install_directory)

    var f:FileAccess = FileAccess.open(download_path,FileAccess.WRITE)
    f.store_buffer(body)
    f = null
    unzip()

func unzip():
    var reader = ZIPReader.new()
    var err = reader.open(download_path)
    print(err)
    if err != OK:
        push_error('An error occured reading the zip file')
        return err
    else:
        print("I'm good")
```
Button 3 moves the ZIPReader part back to _on_pressed by using await and a signal, emitted after the file is written
```gdscript
extends Button

var download_path = 'down/zip.zip'
var download_link = 'https://github.com/oganm/ZIPReader_issue/archive/refs/heads/main.zip'
signal zip_written

func _ready():
    pressed.connect(self._on_pressed)

func _on_pressed():
    DirAccess.remove_absolute(download_path)

    var http_request = HTTPRequest.new()
    add_child(http_request)
    http_request.request_completed.connect(self._download_completed)
    http_request.request(download_link)
    await self.zip_written

    var reader = ZIPReader.new()
    var err = reader.open(download_path)
    print(err)
    if err != OK:
        push_error('An error occured reading the zip file')
        return err
    else:
        print("I'm good")

func _download_completed(result, response_code, headers, body):
    var install_directory = download_path.get_base_dir()
    DirAccess.make_dir_recursive_absolute(install_directory)

    var f:FileAccess = FileAccess.open(download_path,FileAccess.WRITE)
    f.store_buffer(body)
    f = null
    emit_signal('zip_written')
```

### Steps to reproduce

Running the project attached, only pressing button 1 works to successfully read the downloaded zip file.

### Minimal reproduction project

https://github.com/oganm/ZIPReader_issue/archive/refs/heads/main.zip
https://github.com/oganm/ZIPReader_issue

### LINKED ISSUE #69504: GDScript Assigning to a variable 'self' using a cast of type 'RefCounted' increases reference count twice
### Godot version

4.0 beta6

### System information

Windows 10

### Issue description

When getting a reference to 'self' using a cast of type 'RefCounted' increases the reference counter twice instead of once.

Is this an Issue or a misconception around the casting mechanism ?
Should I consider other ways to save an Array of heterogeneous collection of RefCounted subclasses ?

### Steps to reproduce

Instance the following class: (Or just run the supplied project main scene)
```gdscript
extends RefCounted
class_name ReferenceIssueClass

func _init()-> void:
  # Showcase Expected behaviour with and without cast on other objects
  # This section prints 2 then 3
  var otherstuff = ConfigFile.new()
  print(otherstuff.get_reference_count())
  var ref_other : RefCounted = otherstuff as RefCounted
  print(otherstuff.get_reference_count())
  
  # This section prints 2 then 3
  var otherstuff2 = ConfigFile.new()
  print(otherstuff2.get_reference_count())
  var ref_other2 : RefCounted = otherstuff2
  print(otherstuff2.get_reference_count())
  
  # Expected behaviour on self without cast
  # This section prints 2 then 3
  print(self.get_reference_count())
  var ref_nocast : RefCounted = self
  print(self.get_reference_count())
  
  # Weird behaviour on self with cast
  # This section prints 3 then 5
  print(self.get_reference_count())
  var ref_cast : RefCounted = self as RefCounted
  print(self.get_reference_count())
```
Prints to console:
```
2
3
2
3
2
3
3
5 # Should be 4 ?
```

### Minimal reproduction project

[selfreferenceissue.zip](https://github.com/godotengine/godot/files/10143382/selfreferenceissue.zip)

### LINKED ISSUE #69551: FileAccess lock
### Godot version

v4.0.beta7.official.0bb1e89fb

### System information

Windows 10 64 bits Vulkan API 1.2.0 

### Issue description

Relates to FileAccess. Attempts to re-read data from file with get_var() immediately after having saved it with store_var() fails (returns null). However, if the same operation is delayed without any change to the code, it works. 

This led me to believe that the internal unlocking of the file takes time despite the fact that I assign null to the FileAccess handle before reopening the file since the function file.close() no longer exists.

The data saved is a smallish array of dictionaries (saved size about 3k). I checked the save file with an hex editor and all seems well.

### Steps to reproduce

Create an array of dictionaries
save it to file using store_var()
close the file by assigning null to the FileAccess handle
reopen the file (for reading)
load the saved data with get_var()
it returns null
Repeat the same operation delaying reopening the file by using a Timer and it works

### Minimal reproduction project

The issue is too basic to include code or a project. If needed I will supply the code.

### LINKED ISSUE #73311: DirAccess.rename() fails with files opened in same function
### Godot version

v4.0.rc2.official [d2699dc7a]

### System information

Windows 10

### Issue description

If a file has been previously opened in the same function, it can't be renamed ...sometimes. The behavior differs from 3.5.1, where renaming works as long as the file is closed.
3.5.1: works
```gdscript
var dir: Directory = Directory.new()
if dir.open("res://") != OK:
	return
var file: File = File.new()
file.open("res://test.sav.temp", File.WRITE)
file.close()
dir.rename("res://test.sav.temp", "res://test.sav")
```
rc2: test1 fails, but test2 works
```gdscript
func test1() -> void:
	var dir_access: DirAccess = DirAccess.open("res://")
	var file_access: FileAccess = FileAccess.open("res://failed1.txt", FileAccess.WRITE)
	file_access = null
	dir_access.rename("failed1.txt", "ok1.txt") # FAILS

func test2() -> void:
	var file_access: FileAccess = FileAccess.open("res://failed2.txt", FileAccess.WRITE)
	file_access = null
	var dir_access: DirAccess = DirAccess.open("res://")
	dir_access.rename("failed2.txt", "ok2.txt") # OK
```

See reproduction section for different cases where the rename works and fails.

### Steps to reproduce

```gdscript
extends Node

func _ready() -> void:
	test1() # FAIL
	test2() # OK
	test3() # OK
	test4() # FAIL
	test5() # OK
	test6() # FAIL

func test1() -> void:
	var dir_access: DirAccess = DirAccess.open("res://")
	var file_access: FileAccess = FileAccess.open("res://failed1.txt", FileAccess.WRITE)
	file_access = null
	dir_access.rename("failed1.txt", "ok1.txt")

func test2() -> void:
	var file_access: FileAccess = FileAccess.open("res://failed2.txt", FileAccess.WRITE)
	file_access = null
	var dir_access: DirAccess = DirAccess.open("res://")
	dir_access.rename("failed2.txt", "ok2.txt")

func test3() -> void:
	(func (): var file_access: FileAccess = FileAccess.open("res://failed3.txt", FileAccess.WRITE)).call()
	DirAccess.rename_absolute("res://failed3.txt", "res://ok3.txt")

func test4() -> void:
	if true:
		var file_access: FileAccess = FileAccess.open("res://failed4.txt", FileAccess.WRITE)
	DirAccess.rename_absolute("res://failed4.txt", "res://ok4.txt")

func test5() -> void:
	if true:
		var file_access: FileAccess = FileAccess.open("res://failed5.txt", FileAccess.WRITE)
		file_access = null
	await get_tree().process_frame
	DirAccess.rename_absolute("res://failed5.txt", "res://ok5.txt")

func test6() -> void:
	var file_access: FileAccess = FileAccess.open("res://failed6.txt", FileAccess.WRITE)
	file_access = null
	DirAccess.rename_absolute("res://failed6.txt", "res://ok6.txt")
```

### Minimal reproduction project

N/A

### LINKED ISSUE #73355: GDScript FileAccess reference scope bug
### Godot version

4.0 dev d2b1474da79a4dce5c2031b3a3fafe8aaa2a161f

### System information

Windows 10

### Issue description

It seems like scoped reference objects like FileAccess sometimes live longer than they should, leading to all kind of strange behaviors. Mostly related to file handling encountering weird causality issues, like seeing temporary files that should no longer exist, files not existing even though you just wrote to them, reading back partially written files even though they should be already closed etc. Needless to say, these can cause subtle bugs and corrupted data in tool scripts. These are some reports that I suspect are caused by this bug:

* https://github.com/godotengine/godot/issues/73311
* https://github.com/godotengine/godot/issues/69551
* https://github.com/godotengine/godot/issues/68614
* https://github.com/godotengine/godot/issues/66818

This issue probably also applies to other classes than FileAccess, but it's the easiest place to see these things happen especially as the files are now closed by dropping all references to it instead of calling .close() which makes this bug easy to encounter.

In short, code like:

``` gdscript
var file = FileAccess.open("res://somefile.something", FileAccess.WRITE)
...
file = null
```

does not always free the reference and close the file, sometimes it happens only once the function/variable scope ends.

### Steps to reproduce

Open the script editor and run test_editor_script.gd (right click -> Run). Observe output. Uncomment test_2_ok(), run again and observe correct output. You can also run the project which calls the same code.

Make sure "filesystem/on_save/safe_save_on_backup_then_rename" is enabled (in Editor -> Editor Settings... -> FileSystem -> On Save). Then restart editor. Enabling temp files in not required, it just makes the bug easier to see in action. The test projects prints these temp files so the issue is easier to see as the temp files are renamed to overwrite the real file when the FileAccess object is actually closed. I guess this might affect a lot of scoped reference objects, but it's probably easiest to see when it manifests in external resources like files.

Optionally, apply this patch to the Godot source code C++ file access and rebuild the engine. It simply logs when a file with a certain type of name is actually closed. Note that this one is only for Windows FileAccess, each platform has its own.

```
@@ -155,10 +155,14 @@ Error FileAccessWindows::open_internal(const String &p_path, int p_mode_flags) {
 void FileAccessWindows::_close() {
 	if (!f) {
 		return;
 	}
 
+	if (path.find("bugfile") != -1) {
+		print_line("Close file: " + path);
+	}
+
 	fclose(f);
 	f = nullptr;
 
 	if (!save_path.is_empty()) {
 		bool rename_error = true;
```

Then run the test project.

``` gdscript
func test_fails():
    var file = FileAccess.open("res://bugfile_1.noextension", FileAccess.WRITE)
    var file2 = FileAccess.open("res://bugfile_2.noextension", FileAccess.WRITE)

    #These do not close both files, patch FileAccess C++ code to log file close calls.
    #Only first file is closed, file2 is closed only after function scope ends.
    file = null
    file2 = null

    print("The End")
```

With the patched engine this produces output

```
Close file: E:/Repositories/projects/file-close-bug/bugfile_1.noextension.tmp
The End
Close file: E:/Repositories/projects/file-close-bug/bugfile_2.noextension.tmp
```

Huh. Note that file2 is closed only after the print("The End") and once the function scope ends. Simply adding a junk reference counted variable "fixes" this:

``` gdscript
func test():
    var file = FileAccess.open("res://bugfile_1.noextension", FileAccess.WRITE)
    var file2 = FileAccess.open("res://bugfile_2.noextension", FileAccess.WRITE)

    #Simply adding this will close both files correctly and cleans up temp files
    #before function scope ends. Moving this before file creation will restore buggy file2 close behavior.
    var useless = RefCounted.new()

    file = null
    file2 = null

    print("The End")
```

This produces correct output:

```
Close file: E:/Repositories/projects/file-close-bug/bugfile_1.noextension.tmp
Close file: E:/Repositories/projects/file-close-bug/bugfile_2.noextension.tmp
The End
```

Looks like the last reference counted object in the stack has some lifetime issues or something similar? Using an int or string instead of RefCounted doesn't seem to work to "fix" the second test. It's kind of hard to say what is the root cause for this behavior. At first I was even thinking maybe print()'s were just buffered in wrong order somehow, but in the test project you can see the behavior with temp files, too.

### Minimal reproduction project

[file-close-bug.zip](https://github.com/godotengine/godot/files/10743162/file-close-bug.zip)

### PR DESCRIPTION
**Pull request by RandomShaper**
I could have added a new opcode hardwired to assign null, which looks kind of more natural for the task, but assigning `false` is fine, since there's already an opcode for that, and any builtin type would do as well as a null object.

This has a potential for optimization; i.e., there's no real need to clean up a temporary that is going to be reused immediately, but I guess that can come in a future iteration. I'd be surprised if this implementation hurted performance badly.

**UPDATE:** Consolidated the list of issues fixed.
Confirmed fixed the following (at least, MRPs work fine):
- Fixes #73355.
- Fixes #73311.
- Fixes #69551.
- Fixes #68614.
- Fixes #66818 (see comment there).
- Fixes #69504 (whether it's a bug it's a bit blurry to me, but with this PR the MRP works as expected).

**NOTE:** Forgot to say that I haven't tested this extensively. It'd be nice that someone run a more or less script intensive project to hunt down issues.

### CODE REVIEW

**Comment by RandomShaper** [comment]
Finished testing all the related issues. For what it's worth, it seems I haven't broken GDScript with this, but this PR touches a critical aspect of Godot and I don't want to break RC 3. 😅 I hope someone watching can run a couple of real world projects or examples to confirm this is good.

**Comment by Lielay9** [comment]
> I hope someone watching can run a couple of real world projects or examples to confirm this is good.

The project #73311 was extracted from seems to work without exploding immediately and the cases regarding `FileAccess` lingering longer than expected seem to be fixed.

**Comment by RandomShaper** [comment]
> This will relatively significantly impact the performance of GDScript, right? Anytime a temporary address is used to store an object, the VM will execute another instruction to clear it from the stack?
> 
> I suspect this might just also fix #69504, would you be able to check?

Some impact is expected, but it's hard to know how much in advance. The `Variant` being "disarmed" is likely in cache and the operation is potentially very fast, though not free. Only measuring one can tell for sure how much it slows things down.

I'll check the other issue tomorrow.

**Comment by anvilfolk** [comment]
> I'll check the other issue tomorrow.

The reason I brought that argument up was because performance was partly the reason we decided to let the other issue be! We figured there was no problem with allowing some things to live in the stack a little longer than they would otherwise - but that clearly has a bunch of other nasty side-effects which look like they will be fixed by this PR, so that's super exciting!

**Comment by dalexeev** [comment]
I tested it, found a regression. Minimal example to reproduce:

```gdscript
@tool
extends EditorScript

var nodes := {}

func test(name: String) -> void:
    var node: Node = nodes.get(name) # Error: Trying to assign invalid null variable.
    if !node:
        node = Node.new()
        nodes[name] = nodes
    node.print_tree()

func _run() -> void:
    test("test")
```

**Comment by RandomShaper** [comment]
Checked #69504. With this PR, the expectation in the MRP holds. Updating list in the description.

**Comment by RandomShaper** [comment]
> The reason I brought that argument up was because performance was partly the reason we decided to let the other issue be!

I'm pretty sure there's a way to avoid most of the extra instructions, in a way that the RC is not artificially incremented without having to mechanically clean each and every case like this PR does. However, for now and given I can't spend much more time on this, I'd consider this patch to be already an improvement.

File: modules/gdscript/gdscript_byte_codegen.cpp [hunk 1]
```diff
@@ -143,6 +143,14 @@ void GDScriptByteCodeGenerator::pop_temporary() {
 	ERR_FAIL_COND(used_temporaries.is_empty());
 	int slot_idx = used_temporaries.back()->get();
 	const StackSlot &slot = temporaries[slot_idx];
+	if (slot.type == Variant::OBJECT) {
+		// Avoid keeping in the stack long-lived references to objects,
+		// which may prevent RefCounted objects from being freed.
+		GDScriptDataType type = {};
+		type.kind = GDScriptDataType::BUILTIN;
+		type.builtin_type = slot.type;
```
Comment (vnen) [inline modules/gdscript/gdscript_byte_codegen.cpp:151]: Setting the type here is not needed. This won't be used by `write_assign_false()`.

Comment (RandomShaper) [reply modules/gdscript/gdscript_byte_codegen.cpp:151]: So can I just do `write_assign_false(Address(Address::TEMPORARY, slot_idx));`?

Comment (vnen) [reply modules/gdscript/gdscript_byte_codegen.cpp:151]: Yes.

**Comment by vnen** [comment]
So, I think this a valid solution. The only problem is the `get_call_target()` function:
https://github.com/godotengine/godot/blob/f84479fe5a0a3c34c1d0217527ae08dd24543243/modules/gdscript/gdscript_byte_codegen.cpp#L957-L971

It operates under the assumption that `pop_temporary()` does nothing special and perform a preemptive pop. This avoids having to do it on every place it is called. However, with the changes in this PR this assumption no longer holds. If it `get_call_target()` happens to find an Object, it will reach the condition added here and mangle the bytecode (since it will emit a instruction in the middle of emitting another).

For this PR to work we'll need to change how `get_call_target()` works. I think a solution for this is to create a wrapper struct that it can return which holds the temporary address. It knows whether the address is a new temporary or not, so it can be responsible from popping the temporary in the destructor. Then we can use RAII to not have to remember to call `pop_temporary()` every time `get_call_target()` (also, this should only be done **if** `get_call_target()` created the temporary, which is not obvious for the callers).

Something like this:
```cpp
struct CallTarget {
		Address target;
		bool is_new_temporary = false;
		GDScriptByteCodeGenerator *codegen = nullptr;
		CallTarget(Address p_target, bool p_is_new_temporary, GDScriptByteCodeGenerator *p_codegen) :
				target(p_target),
				is_new_temporary(p_is_new_temporary),
				codegen(p_codegen) {}
		~CallTarget() {
			if (is_new_temporary) {
				codegen->pop_temporary();
			}
		}
	};
```

Then, naturally, all calls to `get_call_target()` will need to be changed to use this.

Not sure if @RandomShaper follows my thought, given the unfamiliarity with this part of code. If you want I can implement this bit.

**Comment by RandomShaper** [comment]
@vnen, I think I've done it. I've diverted from your idea in a single thing: cleanup is explicit, for two reasons:
- Enabling RAII would require the `CallTarget` class to have some sort of ref counting.
- This way makes the code very explicit about the cleanup and also gives a chance for checking for misuse.

I've obvsiously done something wrong since the GDScript test suite is crashing. I'm trying to debug it.

**Comment by RandomShaper** [comment]
I've fixed a silly error and it seems to work. At least, I've tested it with a few script-intensive Godot demo projects and they seemed to work fine.

**Review by vnen — APPROVED** [review]
Looks good to me. Thanks for working on this!

**Comment by YuriSizov** [comment]
Let's give it a good over-the-weekend testing :)
